http: `to_reply()` from `redirect_exception`

WillemKauf opened the pull request:
This PR adds `redirect_exception::to_reply()` for construction of a `http::reply` from a redirect case, with a `Location` header and optionally a `Retry-After` header.

xemul commented:
@nyh , please review

nyh commented on include/seastar/http/exception.hh:
```diff
@@ -65,10 +65,21 @@ private:
-    redirect_exception(const std::string& url, http::reply::status_type status = http::reply::status_type::moved_permanently)
-            : base_exception("", status), url(url) {
+    redirect_exception(const std::string& url, http::reply::status_type status = http::reply::status_type::moved_permanently, const std::optional<int>& retry_after = std::nullopt)
+            : base_exception("", status), url(url), retry_after(retry_after) {
```
You should mention in the commit message (and perhaps even split the commit into two) that one of the things that this patch does is to add to add Retry-After header (see https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.3) support for redirections.
It is separate feature from the to_reply() function you are adding in this patch.

nyh commented on src/http/routes.cc:
```diff
@@ -94,6 +93,9 @@ future<std::unique_ptr<http::reply> > routes::handle(const sstring& path, std::u
             return r.handle_exception(_general_handler);
+        } catch (const redirect_exception& _e) {
+            *rep = _e.to_reply();
+            rep->done("json");
```
This catch was missing previously. It looks like it fixes a bug - can you explain in the commit message what bug this was? Can you perhaps write a unit test for it?

nyh commented:
The commit message made it sound like your patch only added a new API function, but as I noted in the review above, it actually included two more separate changes:

1. It adds support for Retry-After (https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.3), at least the part for 3xx (redirection), which was missing before. However, I want to ask something (because my command of the RFC is quite rusty) Is this the really only header that makes sense to put on a redirection reply? Shouldn't a redirection allow a more general set of headers to be given, and not specifically Retry-After (which isn't a particularly useful or commonly-used header, to be honest).
2. It adds a missing catch of redirect_exception. I am guessing there was a reason to adding this, and it's a bug fixed, but it needs to be explained and justified

travisdowns committed:
http: add to_reply() to redirect_exception with extra-header support

Add redirect_exception::to_reply() which constructs an http::reply
with the Location header and status code pre-populated from the
exception.  An optional initializer-list of extra headers (e.g.
Retry-After, as per RFC 7231 §7.1.3) may be passed to the constructor
and are forwarded to the reply unchanged, giving callers full control
over the response headers without requiring a more specialised API.

Use to_reply() in routes::exception_reply() so that the existing
redirect path also honours any extra headers.

Also add an explicit catch(redirect_exception) in routes::handle() to
cover the case where a handler throws synchronously (before returning a
future).  In that path rep has been moved into the handler, so the
catch creates a fresh unique_ptr rather than dereferencing the
now-null one.  Without this catch the synchronous-throw case fell
through to the catch-all which called exception_reply(); both paths
produce equivalent output, but being explicit avoids the latent
nullptr dereference and ensures consistency with the async path.

Rebased from Willem Kaufmann <willem.kaufmann@redpanda.com> (6196a6823).

WillemKauf commented:
> Apologies that this sat unaddressed for so long — I'm helping move it along.

Yes, thank you Travis! :) sorry I let this rot.

nyh commented on src/http/routes.cc:
```diff
@@ -96,6 +95,10 @@ future<std::unique_ptr<http::reply> > routes::handle(const sstring& path, std::u
+        } catch (const redirect_exception& _e) {
+            // rep may have been moved into handler->handle(), so create a fresh one
+            rep = std::make_unique<http::reply>(_e.to_reply());
+            rep->set_content_type("json");
```
I see this set_content_type("json") is preexisting, so it's ok that you did it too, but I wonder who/why decided that Seastar's error messages should be in JSON. I think it should have been the application's choice.

But in any case, I am guessing most existing applications already bypass this code, as it's really not general enough.
